### PR TITLE
[PLAT-862] Storage usage for osfstorage

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Let the user know how much storage they are using.

## Changes

- add property to Osfstorage NodeSettings that sums the size of each every version with test.
- minor frontend changes to make file size visible. 

## QA Notes

To observe these changes upload different files and different versions to the files tree. Reload the page. Their should be a number next to the OSF storage label that represents the storage usage of all files and all versions at that node. like the following:

screen shot 2018-06-05 at 4 54 58 pm

## Documentation

I'm unsure if this needs more context, it doesn't specify anything extra in the ticket.

## Side Effects

This feature is waffled under flag 'storage_usage'.

## Ticket

https://openscience.atlassian.net/projects/PLAT/issues/PLAT-862

### CODE REVIEW

File: addons/osfstorage/models.py [hunk 1]
```diff
@@ -536,3 +536,14 @@ def create_waterbutler_log(self, auth, action, metadata):
             auth=auth,
             params=params
         )
+
+    @property
```
Comment (erinspace) [inline addons/osfstorage/models.py]: You'd have to confirm with @sloria, but this might be a good candidate for `cached_property` as  I imagine this'll be an expensive property, but I dunno if it has any unintended side effects...

Also, does this PR deal with the bullet point in the dev notes on this ticket "Cache the total until there is a change (file uploaded, file deleted, etc.)"?

Comment (erinspace) [reply addons/osfstorage/models.py]: Somewhat unrelated, but it looks like totals for linked nodes are getting calculated and displayed on the same page as the node you're on currently -- I know this doesn't count towards the "total," but it might be un-necessarily expensive if we don't care about that total here. Doesn't seem to be too much of a problem except maybe in really really big linked nodes I'm not sure!

screen shot 2018-06-26 at 4 41 55 pm

Comment (Johnetordoff) [reply addons/osfstorage/models.py]: I didn't do any bench-marking, I just since we don't really have to recurse through anything and we're doing the summing in raw sql it should just be fast enough.

A for linked nodes the ticket says they shouldn't be included in the total, but I assume they are still appear like component nodes.

**Comment by sloria** [comment]
OK, discussed a bit more with Michael.

To address the performance issue, I suggest using Django's built-in caching framework to cache the result of the queries for some time (1 min?). For now, we can use the database backend: https://docs.djangoproject.com/en/1.11/topics/cache/#database-caching .

In the future, we could store the file version sizes in Elasticsearch (using django-elasticsearch-metrics). That would allow us to track storage usage over time, internally. But that doesn't need to happen before this feature is released.

File: addons/osfstorage/models.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 from django.apps import apps
 from django.db import models, connection
+from django.core.cache import cache
```
Comment (sloria) [inline addons/osfstorage/models.py]: You'll need to configure the caching backend, so that cache results get stored in the DB. See the docs: https://docs.djangoproject.com/en/1.11/topics/cache/

File: addons/osfstorage/models.py [hunk 3]
```diff
@@ -560,3 +561,20 @@ def create_waterbutler_log(self, auth, action, metadata):
...
+            return storage_usage
+
+        file_versions = OsfStorageFile.objects.filter(provider='osfstorage',
+                                                   target_object_id=self.owner_id,
+                                                   versions__isnull=False).values_list('versions',
+                                                                                       flat=True)
+        # An empty queryset will produce a None instead of 0
+        sum = FileVersion.objects.filter(id__in=file_versions).aggregate(sum=models.Sum('size'))['sum'] or 0
+
+        cache.set('storage_usage', sum, 60)
```
Comment (sloria) [inline addons/osfstorage/models.py]: You'll need to cache the values per-node. Otherwise, we'd be return the most recently-computed value for *every node*. To do this, come up with a unique cache key for each node, e.g. `storage_usage:123` (where 123 is the PK of the node).

File: addons/osfstorage/models.py [hunk 4]
```diff
@@ -560,3 +561,20 @@ def create_waterbutler_log(self, auth, action, metadata):
...
+
+        if storage_usage:
+            return storage_usage
+
+        file_versions = OsfStorageFile.objects.filter(provider='osfstorage',
+                                                   target_object_id=self.owner_id,
+                                                   versions__isnull=False).values_list('versions',
+                                                                                       flat=True)
+        # An empty queryset will produce a None instead of 0
+        sum = FileVersion.objects.filter(id__in=file_versions).aggregate(sum=models.Sum('size'))['sum'] or 0
```
Comment (sloria) [inline addons/osfstorage/models.py]: The cache isn't being used--this query is being called no matter what, which defeats the purpose of caching.

You'll need to

1. Try to get the cached storage usage value for the node
2. If it doesn't exist (or it's expired), then run the query and cache the value for 60 seconds.
3. If it does exist, return that value.

File: addons/osfstorage/models.py (hunk 3 above)
Comment (sloria) [reply addons/osfstorage/models.py]: Also, a nit: `60` is a magic value. Better to set it as a constant.


```python
cache.set(cache_key, value, self.STORAGE_USAGE_CACHE_TIMEOUT)
```

File: addons/osfstorage/models.py (hunk 4 above)
Comment (sloria) [reply addons/osfstorage/models.py]: Also, a nit: `sum` is a reserved word. How about `value` or `total`?

Comment (Johnetordoff) [reply addons/osfstorage/models.py]: I'm confused by this because I'm pretty sure I am actually this is working as intended (aside from the cacheing on a per-node basis). If the value has been cached the method exits early [here](https://github.com/CenterForOpenScience/osf.io/pull/8414/files/00c56995d72fad3e1cf9cab0b7754ad5966ebd0d#diff-511d556f4513d9e41850aa51126eceabR570). otherwise the query runs. I even have to [bust the cache](https://github.com/CenterForOpenScience/osf.io/pull/8414/files/00c56995d72fad3e1cf9cab0b7754ad5966ebd0d#diff-df93ad330ce69a1ef2564e15f3ac0bd6R636) for the unit test to work

**Comment by sloria** [comment]
Also, can you put this behind a waffle feature flag? I'm still concerned about how this will perform on prod. It would be great if we could do a phased rollout on prod. For example, we could enable it for team members with large projects (Brian N). Once we verify it's OK, then we'd enable it for everyone.

File: website/settings/defaults.py [hunk 5]
```diff
@@ -1934,3 +1934,10 @@ class CeleryConfig:
     'googleGroup': 'https://groups.google.com/forum/#!forum/openscienceframework',
     'github': 'https://www.github.com/centerforopenscience',
 }
+
+CACHES = {
+    'default': {
+        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
```
Comment (sloria) [inline website/settings/defaults.py]: Drive-by: We don't have memcached set up. We'll need to use the Django database backend for now.

Comment (sloria) [reply website/settings/defaults.py]: I mean we don't have the [memcached](https://www.memcached.org/) service deployed.

So rather than relying on new infrastructure, we should just use our existing DB for now.

File: addons/osfstorage/models.py [hunk 6]
```diff
@@ -560,3 +564,20 @@ def create_waterbutler_log(self, auth, action, metadata):
             auth=auth,
             params=params
         )
+
+    @property
+    def storage_usage(self):
+        storage_usage = cache.get('storage_usage' + self._id)
```
Comment (sloria) [inline addons/osfstorage/models.py]: Drive-by: we should probably separate the prefix from the variable part, e.g. `storage_usage:123ab` rather than `storage_usage123ab`. This will allow us to user `str.split` to get the GUID, if need be.

File: api/base/settings/defaults.py [hunk 7]
```diff
@@ -284,3 +284,16 @@
 
 # If set to True, automated tests with extra queries will fail.
 NPLUSONE_RAISE = False
+
+CACHES = {
+    'default': {
+        'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
+        'LOCATION': 'osf_cache_table',
+    },
+    'waffle_cache': {
```
Comment (sloria) [inline api/base/settings/defaults.py]: Does this change the behavior in waffle in any way? Or was this added to preserve old behavior?

File: osf/migrations/0129_create_cache_table.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0128_merge_20180829_0012'),
+    ]
+
+    operations = [
+        migrations.RunSQL([
```
Comment (sloria) [inline osf/migrations/0129_create_cache_table.py]: Is this migration necessary? There's a command for this: `python manage.py createcachetable`. Will Django fail gracefully if the database table doesn't exist? If so, then I think this can be removed.

Comment (Johnetordoff) [reply osf/migrations/0129_create_cache_table.py]: Django doesn't fail gracefully essentially it acts like there's an unrun migration.

File: api/base/settings/defaults.py (hunk 7 above)
Comment (Johnetordoff) [reply api/base/settings/defaults.py]: There's a Travis bug if waffle can't cache normally. My understanding of it is running waffle migrations before the cachetable migration means that waffle tries to cache in the DB table which doesn't exist yet. Here's an example: https://travis-ci.org/Johnetordoff/osf.io/jobs/427258977

File: osf/migrations/0129_create_cache_table.py (hunk 8 above)
Comment (sloria) [reply osf/migrations/0129_create_cache_table.py]: >  it acts like there's an unrun migration.

Does that mean that 1) the app doesn't start or 2) an error gets raised when calling `cache.get`? If it's the former, that's good because the app containers won't fail over, which will give us a chance to run the `createcachetable` command.

File: api/base/settings/defaults.py (hunk 7 above)
Comment (sloria) [reply api/base/settings/defaults.py]: Oh ok. But is waffle caching in the same way (LocalMemCache) that it was before this change?

File: osf/migrations/0129_create_cache_table.py (hunk 8 above)
Comment (Johnetordoff) [reply osf/migrations/0129_create_cache_table.py]: It fails when you call `cache.get`. Devs will get an opportunity to run `createcachetable`

File: api/base/settings/defaults.py (hunk 7 above)
Comment (Johnetordoff) [reply api/base/settings/defaults.py]: Yep

Comment (sloria) [reply api/base/settings/defaults.py]: OK, great. 👌

File: osf/migrations/0129_create_cache_table.py (hunk 8 above)
Comment (sloria) [reply osf/migrations/0129_create_cache_table.py]: Ah, ok. Then I think this migration can stay. Just one suggestion: it should read from `settings` to get the table name. That way, if that setting changes, this migration won't create the wrong table.

**Comment by sloria** [comment]
Looks like this just needs a merge migration (or "rebase", i.e. just remake the migration at the end).

File: website/util/rubeus.py [hunk 9]
```diff
@@ -116,6 +118,11 @@ def build_addon_root(node_settings, name, permissions=None,
         'nodeApiUrl': node_settings.owner.api_url,
     }
     ret.update(kwargs)
+
+    if flag_is_active(request, 'storage_usage_feature'):
```
Comment (sloria) [inline website/util/rubeus.py]: This won't quite work because you're passing a Flask request and `waffle.flag_is_active`. See `website/ember_osf_web/decorators.py` for the workaround we did for storage_i18n.

**Comment by sloria** [comment]
Also, how difficult would it be to update the cache value when a file is added or removed? It looks like we already have a signal that gets triggered when the waterbutler callback happens:

```python
@file_signals.file_updated.connect
def addon_delete_file_node(self, target, user, event_type, payload):
```

Ideally, the user should immediately see the correct value after adding or removing a file.

**Comment by Johnetordoff** [comment]
@sloria The difficulty in seeing the correct value immediately comes more from Fangorn then from the backend. When upload or delete a file you only get info about that file in the response. I'm not sure how we get addon info to fangorn.

**Comment by sloria** [comment]
> The difficulty in seeing the correct value immediately comes more from Fangorn then from the backend. When upload or delete a file you only get info about that file in the response. I'm not sure how we get addon info to fangorn.

Why is it related to fangorn?

To clarify my suggestion: you should be able to update the cached storage value for a node when a file is added or removed. In pseudocode:

```python
@file_signals.file_updated.connect
def addon_delete_file_node(self, target, user, event_type, payload):
   if event_type == 'file_added':
        get cached storage value for target
        increment the storage value by the size of the file
    elif event_type == 'file_deleted':
      get cached storage value for target
      decrement the storage value by the size of the file
```

**Comment by Johnetordoff** [comment]
Oh I thought you meant the user would see the value change immediately on upload/delete as a opposed to on refresh. We could the above to sync the cache size.

**Comment by Johnetordoff** [comment]
@sloria I have solution here which updates the cache more dynamically, however I'm not sure this is actually faster for most projects. We have to do some recursive querying for when folders are deleted. I can probably optimize this further, but if I'm unsure if it's worth it considering how much work we're doing to recover the size metadata.

File: tests/test_addons.py [hunk 10]
```diff
@@ -366,18 +375,27 @@ def test_action_downloads_contrib(self):
...
         url = self.node.api_url_for('create_waterbutler_log')
-        payload = self.build_payload(metadata={'materialized': path, 'kind': 'file', 'path': path})
+        payload = self.build_payload(metadata={'materialized': self.file.materialized_path, 'kind': 'file', 'path': self.file.path})
         nlogs = self.node.logs.count()
         self.app.put_json(url, payload, headers={'Content-Type': 'application/json'})
-        self.node.reload()
-        assert_equal(self.node.logs.count(), nlogs + 1)
-        assert('urls' in self.node.logs.filter(action='osf_storage_file_added')[0].params)
+
+    @mock.patch('addons.base.views.storage_usage_flag_active', return_value=True)
```
Comment (sloria) [inline tests/test_addons.py]: `waffle.testutils.override_flag` is the preferred way to override flags.

File: osf/migrations/0133_merge_20180918_1436.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.13 on 2018-09-18 14:36
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
Comment (sloria) [inline osf/migrations/0133_merge_20180918_1436.py]: Not sure why there were two merge migrations created. Would you mind updating with develop and "rebasing" the migration (i.e. putting it at the end)?

File: addons/base/views.py [hunk 12]
```diff
@@ -532,6 +534,44 @@ def addon_delete_file_node(self, target, user, event_type, payload):
...
+
+            if event_type == 'file_removed':
+                if path.endswith('/'):
+                    folder = TrashedFolder.objects.get(_id=path.rstrip('/'))
+                    storage_usage -= get_storage_usage_of_children(folder)
+                else:
+                    trashed_node = TrashedFile.objects.get(_path='/' + path)
+                    storage_usage -= trashed_node.versions.all().aggregate(sum=Sum('size'))['sum']
+
+            cache.set('storage_usage:' + osfstorage._id, storage_usage, osfstorage.STORAGE_USAGE_CACHE_TIMEOUT)
```
Comment (sloria) [inline addons/base/views.py]: Nit: prefer `str.format` to concatenation.

File: addons/base/views.py [hunk 13]
```diff
@@ -532,6 +534,44 @@ def addon_delete_file_node(self, target, user, event_type, payload):
...
+        if provider == 'osfstorage':
+            path = payload['metadata']['path']
+            osfstorage = target.get_addon('osfstorage')
+            storage_usage = cache.get('storage_usage:' + osfstorage._id)
+
+            if not storage_usage:
+                storage_usage = osfstorage.storage_usage
+            elif event_type in ('file_added', 'file_updated'):
+                file_node = OsfStorageFile.objects.get(_id=path)
+                storage_usage += file_node.versions.first().size
```
Comment (sloria) [inline addons/base/views.py]: Is `first()` guaranteed to return the latest version? Better to add an explicit `order_by` to make sure.

File: addons/base/views.py [hunk 14]
```diff
@@ -532,6 +534,44 @@ def addon_delete_file_node(self, target, user, event_type, payload):
...
+
+            if event_type == 'file_removed':
+                if path.endswith('/'):
+                    folder = TrashedFolder.objects.get(_id=path.rstrip('/'))
+                    storage_usage -= get_storage_usage_of_children(folder)
```
Comment (sloria) [inline addons/base/views.py]: I see what you mean about having to recursively sum the sizes of files within a folder.

In order to simplify this, how about you just invalidate the cache key when a folder is deleted?

File: addons/base/views.py (hunk 13 above)
Comment (sloria) [reply addons/base/views.py]: Oh, I see, looks like we set `ordering` on `FileVersion`: https://github.com/CenterForOpenScience/osf.io/blob/ffc0b2cb22a7c08a51e878ccb04357d6f9977912/osf/models/files.py#L807-L808 .

So I guess this is fine. Never hurts to be explicit though.

File: osf/migrations/0130_create_cache_table.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.conf import settings
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0129_merge_20180910_1926'),
```
Comment (sloria) [inline osf/migrations/0130_create_cache_table.py]: You'll need to change this to the latest migration.

File: website/ember_osf_web/decorators.py [hunk 16]
```diff
@@ -33,3 +33,8 @@ def wrapped(*args, **kwargs):
 def storage_i18n_flag_active():
     request.user = _get_current_user() or MockUser()
     return waffle.flag_is_active(request, 'storage_i18n')
+
+
+def storage_usage_flag_active():
+    request.user = _get_current_user() or MockUser()
+    return waffle.flag_is_active(request, 'storage_usage')
```
Comment (sloria) [inline website/ember_osf_web/decorators.py]: Can you please move `'storage_usage'` to a constant in `osf/features.py` and use that here?

File: addons/osfstorage/models.py [hunk 17]
```diff
@@ -591,3 +595,20 @@ def create_waterbutler_log(self, auth, action, metadata):
...
+
+        if storage_usage:
+            return storage_usage
+
+        file_versions = OsfStorageFile.objects.filter(provider='osfstorage',
+                                                   target_object_id=self.owner_id,
+                                                   versions__isnull=False).values_list('versions',
+                                                                                       flat=True)
+        # An empty queryset will produce a None instead of 0
+        storage_usage_total = FileVersion.objects.filter(id__in=file_versions).aggregate(sum=models.Sum('size'))['sum'] or 0
```
Comment (icereval) [inline addons/osfstorage/models.py]: For performance it might be much faster to write this as a single query...

```python
AbstractNode.objects.filter(
    files__type='osf.osfstoragefile',
    id=self.owner_id
).annotate(
    Sum('files__versions__size'),
).values_list(
    'files__versions__size__sum',
).first()
```

Comment (icereval) [reply addons/osfstorage/models.py]: http://tatiyants.com/pev/#/plans/plan_1539805418000
